### PR TITLE
adding huber loss implementation

### PR DESCRIPTION
**Pull request by mrityunjay-tripathi**
Hello Everyone,
I am adding [Huber Loss](https://en.wikipedia.org/wiki/Huber_loss) Function. The main feature of Huber Loss is that it is less sensitive to outliers in data than the Mean Squared Error function.

### CODE REVIEW

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * Computes the Huber Loss function.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param target The target vector.
+   */
+  template<typename InputType, typename TargetType>
+  double Forward(const InputType&& input, const TargetType&& target);
+  /**
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Missing empire line here.

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  //! Set the value of delta.
+  double& Delta() { return delta; }
+
+  //! Get the value of reduction type.
+  bool Mean() const { return mean; }
+  //! Set the value of reduction type.
+  bool& Mean() { return mean; }
+
+  /**
+   * Serialize the layer
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Missing stop here.

**Comment by kartikdutt18** [comment]
Hi @mrityunjay-tripathi , Could you also add something similar [to this](https://github.com/mlpack/mlpack/blob/f68fc14a77fbab24d586092ceac2294f62564b18/src/mlpack/methods/ann/layer/elu.hpp#L263) as smoothL1 loss is huber loss with delta=1, so that it supports both smoothL1 and huber.

**Comment by mrityunjay-tripathi** [comment]
Hello @kartikdutt18
It's not needed as the default value of ```delta``` in Huber loss is 1.
Why to have two functions just for different values of  a hyperparameter?

**Comment by kartikdutt18** [comment]
Hmm, you can leave it then, I didn't see that the default value of delta was 1. I thought of doing so simply because people know more about smoothL1 as it is known bounding box regressor in faster rcnn and since a similar thing was done for selu and elu (for just one different value of hyper parameter i.e. lambda) , I thought we shall also do it here for simplicity for users coming from Pytorch. But I understand if it's not required. Thanks for all your work.

**Comment by mrityunjay-tripathi** [comment]
> Hmm, you can leave it then, I didn't see that the default value of delta was 1. I thought of doing so simply because people know more about smoothL1 as it is known as bounding box regressor in faster rcnn and since a similar thing was done for SELU and ELU (for just one different value of hyper parameter i.e. lambda) , I thought we shall also do it here for simplicity for users coming from Pytorch. If you refer #869 was implemented as ELU only but since SELU as became popular as an activation function even though it just differs by a single hyper parameter, it was added later as a template.
> But I understand if it's not required. Thanks for all your work.

Thanks for the clarification @kartikdutt18 
I think that's where documentation comes into play. We need not make things similar to PyTorch or any other tool. I thought of keeping things concise.

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+>
+class HuberLoss
+{
+ public:
+  /**
+   * Create the HuberLoss object.
+   *
+   * @param delta The threshold value upto which squared error is followed and
+   *              after which absolute error is considered.
+   * @param mean It takes either 1 or 0 i.e. true or false. If true then
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: I'd suggest a light reword here; this description made it sound like an integer type; maybe something like "Boolean. If true, take the mean of total loss. If false, take the sum of total loss."

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
+/**
+ * @file huber_loss.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the Huber Loss function.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: I guess, to be correct we should go with `Definition of the Huber loss function`.

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Huber loss is a loss function used in robust regression,
+ * that is less sensitive to outliers in data than the squared error loss.
+ * This function is quadratic for small values of `y - f(x)`,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: We should use `\f$`, see https://github.com/mlpack/mlpack/blob/93fe75a131c26e422d1c2a65c45cacd4402be6ed/src/mlpack/methods/linear_regression/linear_regression.hpp#L114 for an example. I guess the syntax you used is markdown.

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * and linear for large values, with equal values and slopes of the different
+ * sections at the two points where `|y - f(x)| = delta`.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: See comment above.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @file huber_loss_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Huber Loss function.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: See comment above.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+double HuberLoss<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
+{
+  double lossThis;
+  double totalLoss = 0;
+  double absError;
+
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+      absError = std::abs(target[i] - input[i]);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: I guess you could also use `const double absError = std::abs(target[i] - input[i]);` to make clear the value isn't changed.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      lossThis = absError > delta
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Looks like you could use `totalLoss += absError > delta ...` here.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    const InputType&& input,
+    const TargetType&& target,
+    OutputType&& output)
+{
+  output.set_size(size(input));
+  double absError;
+
+  for (size_t i = 0; i < output.n_elem; ++i)
+  {
+    absError = std::abs(target[i] - input[i]);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Should we use `const double absError` here as well?

File: src/mlpack/tests/loss_functions_test.cpp [hunk 11]
```diff
@@ -33,6 +34,33 @@ using namespace mlpack::ann;
 
 BOOST_AUTO_TEST_SUITE(LossFunctionsTest);
 
+/**
+ * Simple Huber Loss test.
```
Comment (zoq) [inline src/mlpack/tests/loss_functions_test.cpp]: `Huber loss`.

**Comment by mrityunjay-tripathi** [comment]
Thank you @zoq for the review. I was busy with exams. Also, I thought of incorporating the feature suggested by @kartikdutt18 in #2211. I can take-up that for other loss functions as well.

**Comment by kartikdutt18** [comment]
@mrityunjay-tripathi, I don't think we should incorporate changes for #2211 right now, As design for it has not been finalized. For other loss functions, I have already implemented an overloaded function. However, I wanted to send a PR only when I know that it's the best way of doing the same. Once design is finalized, there will be some new loss functions that will be added in mlpack and if you want, we could collaborate on them as I have already made changes for the existing functions. Thanks.

**Comment by birm** [comment]
I feel like #2211 would be better served as its own PR.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+      absError = std::abs(target[i] - input[i]);
+      loss += absError > delta
+          ? delta * (absError - 0.5 * delta) : 0.5 * std::pow(absError, 2);
+  }
+  return mean ? loss / input.n_elem : loss;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void HuberLoss<InputDataType, OutputDataType>::Backward(
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Am I missing something, or do we need to account for mean and sum cases in Backward like in forward?

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: That's correct. I will make the changes as soon as possible.

**Comment by mrityunjay-tripathi** [comment]
Hi @birm, I've made the changes as requested. Can you review the code when you get the time?

**Comment by mrityunjay-tripathi** [comment]
Also, what is the way to cancel the build checks when not needed?? ([CI SKIP] doesn't seem to work here)

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The Huber loss is a loss function used in robust regression,
+ * that is less sensitive to outliers in data than the squared error loss.
+ * This function is quadratic for small values of \f$ y - f(x) \f$,
+ * and linear for large values, with equal values and slopes of the different
+ * sections at the two points where \f$ |y - f(x)| = delta \f$.
+ *
+ * @tparam ActivationFunction Activation function used for the embedding layer.
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Seems like this was included by mistake

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+{
+ public:
+  /**
+   * Create the HuberLoss object.
+   *
+   * @param delta The threshold value upto which squared error is followed and
+   *              after which absolute error is considered.
+   * @param mean If true then mean loss is computed otherwise sum.
+   */
+  HuberLoss(
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: I think this could have the parameter on the same line

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class HuberLoss
+{
+ public:
+  /**
+   * Create the HuberLoss object.
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: `Huber Loss`

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    const InputType& input,
+    const TargetType& target,
+    OutputType& output)
+{
+  output.set_size(size(input));
+  double absError;
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Does this need to be declared outside the loop?

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp (hunk 13 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Ohh yes. It will be removed in the next commit.

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp (hunk 14 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Yup!

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp (hunk 15 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: It's the class name and I guess exact name should be mentioned.

Comment (favre49) [reply src/mlpack/methods/ann/loss_functions/huber_loss.hpp]: Oops, that's my mistake. Sorry about that.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp (hunk 16 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: It can be declared inside the for loop. I will make the modifications in the next commit. Thanks for pointing this out.

**Comment by birm** [comment]
If you merge upstream (mlpack) master into this branch, it should bring back the azure CI tests. Just be careful not to confuse git.

**Comment by mrityunjay-tripathi** [comment]
Hi @birm, I think the build fails have nothing to do with this PR.

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    OutputType& output)
+{
+  output.set_size(size(input));
+  for (size_t i = 0; i < output.n_elem; ++i)
+  {
+    const double absError = std::abs(target[i] - input[i]);
+    output[i] = absError > delta
+        ? - delta * (target[i] - input[i]) / absError : input[i] - target[i];
+    if (mean)
+    {
```
Comment (favre49) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: We can remove the braces here
